appengine-jetty-managed-runtime/src/main/java/com/google/apphosting/runtime/jetty9/SaveSessionFilter.java

```diff
@@ -52,9 +55,25 @@ public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain
     } finally {
       HttpSession session = httpReq.getSession(false);
       if (session instanceof AppEngineSession) {
-        AppEngineSession aeSession = (AppEngineSession) session;
-        if (aeSession.isDirty()) {
-          aeSession.save();
+        final AppEngineSession aeSession = (AppEngineSession) session;
+        
+        if (req.isAsyncStarted())
+        {
+          req.getAsyncContext().addListener(new AsyncListener() {
+            @Override public void onTimeout(AsyncEvent event) throws IOException {}
+            @Override public void onStartAsync(AsyncEvent event) throws IOException {}
+            @Override public void onError(AsyncEvent event) throws IOException {}
+            @Override
+            public void onComplete(AsyncEvent event) throws IOException {  
+              if (aeSession.isDirty()) {
+                aeSession.save();
```

Comment:
@ludoch is this session save behaviour still needed?  Sessions should be able to track how many active requests they have and do a save when this goes to zero.   Saving a session when one request completes may be invalid if another request for the same session is partially completed!